README
// php - netPanzer - Serverbrowser

Contents:
1. Introduction / License / Used/needed libraries
2. Installation / Usage
3. Configuration
4. Contact / who has done this

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1. Introduction
The php-netPanzer-Serverbrowser lets you find netPanzer games on the internet.
Games are displayed in a nice html interface. If you use a newer netPanzer version
(0.8 and newer) on Windows you can yoin a game by klicking the servername.

Php-netPanzer-Serverbrowser is based on the  php-netPanzer-Toolbox which is a
collection of classes and scripts supplying methods to query the masterserver network,
query gameservers, display map images,.... . 

1.1 License
GNU GENERAL PUBLIC LICENSE
See COPYING for more details.

1.2 Used/needed libraries
netPanzer - Browser uses PEAR Cache_Lite to cache an entire dataset. This is necessary
cause of some unreachable (behind firewalls) or crashed servers. The Smarty template
system is used to display the data.

Smarty is included in this release. So you only need to install PEAR Lite_Cache.

Web resources:
PEAR / Cache_Lite: 	http://pear.php.net/package/Cache_Lite
Smarty:				http://smarty.php.net
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2. Installation
Unzip the files contents in any dir on your web server. Make sure the script is allowed
to write in the cache directory. Make sure you have php, PEAR and PEAR Cache_Lite
installed.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2.1 Usage
You use the script by simply calling it with your webbrowser. By default it refreshes
itsself every 45 seconds. The data cache gets refreshed every 60 seconds. Both values
might be changed in the configuration section of index.php .

	Viewing debug output:
If you want to view the debug output, call index.php?raw=1 .

	Doing a manual refresh:
You can manually refresh the server dataset by calling index.php?refresh=1 .
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3. Configuration:
Configuration is divided into to parts. The global configuration in web.conf.php and a
special configuration part in every file. Global configuration gets overriden.

If you have problems understanding the configuration, simply contact me.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4. Contact
php - netPanzer - Browser is programmed by
    Tobias Blersch  ( npb@schrelb.de )

If you want to talk to me via irc, talk to abitos on #netpanzer at irc.freenode.net .


ChangeLog
// netPanzer - game - browser in php - ChangeLog:

9. November 2005 17:39 - Revision: 320
- Added branch php-netpanzer-serverbrowser

9. November 2005 17:20 - Revision: 319
- Commit after several changes.
- Modified templates slightly.
- Solved bug in Filter.php

7. December 2004 21:13 - Revision: 303
 - Added preliminary logparsing functionality.
   - Consists of applications/log.php. This file displays data. Uses several new templates.
   - You can see Top 10 from kills, commander points (kills-losses), onlinetime, matchwins.
   - You can see which tank is best.
   - Logdata is stored in a db -> logdata.sql
 - Added a tankimage output script (applications/gettankimage.php). Data is accuired by a
   class named Tank.php.
 - Added a netpanzer://server:port url to the game browser. Only supported by cvs builds.
   You need to adda regfile to add the netpanzer:// protocol to windows.

27. November 2004 12:05 - Revision: 302
 - Added class Serverliste. The classes Filter and Browser are now derrivated from that file.
 - Added LogAnalysis class. Its derrivated from the LogParser and evalutes its data.
 - Added /applications/cache/logs/ dir. For log-cache data.

24. November 2004 - 20:35 - Revision: 301
 - Many changes!
 - Converted flag images from bmp to png (thanks Bill G.)
 - Added a flag class. It converts flag-IDs to filenames and delivers images.
 - Added a application called getflagimage.php. Delivers flag images.
 - Added sortplayersbyscore option to the filtering class.
 - Added a central config file (base.conf.php). General conf data gets stored in it.
   Specific conf data is still stored in each file. 

20. November 2004 - 17:05 - Revision: 298
 - Changed serverlist filtering. Filtering now is activated in the browsing file.
 - Moved all from the web accessible files to /applications/ .
 - Renamed former index.php to gamebrowser.php
 - Moved cache folder to applications.
 - Added a serverlist file. It displays a list of all avialable maps.

13. November 2004 - 11:45 - Revision: 294 - v0.7
 - Release of Version 0.7

13. November 2004 - 11:45 - Revision: 293
 - Added map spawnpoints information. Code added to Map.php .
 - Added maps .spn files (spawnpoint data).
 - Added a resolvemapinfo filter.
 - Modified README.

12. November 2004 - 20:00 - Revision: 292
 - Changed sorting into filtering. You can sort gameservers, remove dead, full and empty ones.
 - Debugged gameserver sorting.
 - Modified template slightly.

12. November 2004 - 18:00 - Revision: 291
 - Added a sorting - option. Code resides in class Gamesort.php .
 - Modified template. Added info text when servers don't respond. Added a reload button.

12. November 2004 - 16:34 - Revision: 290
 - Had a bug in Map.php (image processing)

12. November 2004 - 16:05 - Revision: 289
 - Moved map raw images to data/maps.
 - Added maps - .opt - files to data/maps.
 - Added class Maps. It processes map information. Images get resized, you can get outpost lists.
 - Modified getimage.php.
 - Added flag images to data/flags. Have to programm a MapID to mapname class.

10. November 2004 - 17:49 - Revision: 286 - v0.6
 - Added file ChangeLog
 - Modified ToDO
 - Mofified index.php, getimage.php . More config - options, added infotext to each file.
 - Modified template slightly.

9. November 2004 - 22:58 - Revision: 285
 - Added map-images
 - Added getimage.php . It displays and resizes / caches images.
 - Modified display.smarty cause of map-image adding.
 - Added TODO.

9. November 2004 - 19:47 - Revision: 283 - v0.5
 - Removed raw.php. You now call raw data using index.php?raw=1
 - Added ability to set timeouts globally for Gameserver, Masterserver, Masterserver.
 - Improved configuration. Now all data resides in array $conf.

2. November 2004 - 15:42 - Revision: 277
 - code got GPL'ed
 - Added README, COPYING file

No log before.

index.php
<?php
/*
Copyright (C) 2004 Tobias Blersch <npb@schrelb.de>
                                                                                
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
                                                                                
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
                                                                                
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* * php - netPanzer - Browser
*	
* Displays all running netPanzer games.
* For  more  info  on  netPanzer  visit
* *  http://netpanzer.berlios.de
*
* * Configuration:
*
* Please note:	  All times are seconds.
*/

/** Include base configuration file: **/
require_once ("base.conf.php");


/** Specific configuration data: **/
$conf = array_merge ($conf, array (
	"cachelifetime" => 	60,				// Time (seconds) after a the script creates a new dataset.
														// Dont keep this value to small. Querying the servers takes quite a lot time.
	"initialmasters" => array (			// Initial list of masterservers.
							"81.169.185.36",
							"81.173.119.122"
						),
	"reloadtime" => 	45,					// Time (seconds) after the webpage gets <meta> refreshed.
	"timeout" => 		1,						// Timeout (seconds) for any queried server.
	"cacheid" => 		"php-netPanzer-Browser",		// Name of the cache item.
	"template" => 		"gamebrowser.smarty",		// Templatefile, ./templates/xy.xy
	"displayconf" => 	1,					// If 1 (true) the configuration is displayed in raw output
													// ( index.php?raw=1).
	"filters" =>		array (				// Applied filters.  Values are descriptive.
							#"notfull",
							#"notempty",
							"notdead",
							"sortbyplayers",
							"sortplayersbykills",
							"sortplayersbyscore",
							"resolvemapinfo",
						)
));

/** Include list with avialable maps **/
require_once ($conf["mapdatadir"] . "maplist.conf.php");

/**
* * Code:
**/

// Include the cache package
require_once('Cache/Lite.php');

// Create a Cache_Lite object
$Cache_Lite = new Cache_Lite(array('cacheDir' => $conf["cachedir"],'lifeTime' => $conf["cachelifetime"]));

if (isset ($_GET["refresh"]) and $_GET["refresh"] == "1") {
	$Cache_Lite->remove ($conf["cacheid"]);
}

// Test if thereis a valide cache for this id
if ($data = $Cache_Lite->get($conf["cacheid"])) {
	$data = unserialize ($data);
	$cachehit = 1;
	// echo "Cache Hit!\n";
} else { // No valid cache found (you have to make the page)
	$cachehit = 0;
	require_once ($conf["includepath"] . "Games/NetPanzer/Browser.php");
	$browser = new Games_NetPanzer_Browser ();
	
	foreach ($conf["initialmasters"] as $initmaster) {
		$browser->addMaster ($initmaster);
	}
	
	$browser->setTimeout ($conf["timeout"]);
	// Resolve game server list.
	$browser->Browse ();
	// Query each server for status info.
	$browser->getGameserversStatus ();
	
	// Apply selected filters.
	require_once ($conf["includepath"] . "Games/NetPanzer/Filter.php");
	$filter = new Games_NetPanzer_Filter ($browser->getGameServers (), $avialablemaps, $conf);
	
	$data["masterservers"] = $browser->getMasterServers();
	$data["gameservers"] = $filter->getGameServers ();
	$data["numgameservers"] = array (
		"all" => $filter->countGameServers (0),
		"active" => $filter->countGameServers ()
	);
    $Cache_Lite->save(serialize ($data));
}

if (!isset ($_GET["raw"]) or $_GET["raw"] != "1") {
	// Smarty output stuff:
	require_once ($conf["smartyclass"]);
	
	$smarty = new Smarty;
	
	$smarty->template_dir = $conf["templatedir"];
	$smarty->compile_dir = $conf["templatecachedir"];
	$smarty->config_dir = './configs/';
	$smarty->cache_dir = './cache/';
	
	$smarty->compile_check = true;
	
	$smarty->assign ("php_self", $_SERVER["PHP_SELF"]);
	$smarty->assign ("masterservers", $data["masterservers"]);
	$smarty->assign ("gameservers", $data["gameservers"]);
	if ($conf["displayconf"] == 1) {
		$smarty->assign ("conf", $conf);
	}
	$smarty->assign ("cachehit", $cachehit);
	$smarty->assign ("numgameservers", $data["numgameservers"]);
	$smarty->display ($conf["template"]);
} else {
	echo "<html><head></head><body><pre>";
	echo "Cache:\n   ";
	if ($cachehit == 1) {
		echo "Hit cache!\n";
	} else {
		echo "I had to build data :-( .\n";
	}
	if ($conf["displayconf"] == 1) {
		echo "\nConfiguration:\n";
		print_r ($conf);
	}
	echo "\nMasterservers:\n";
	print_r ($data["masterservers"]);
	echo "\nGameservers (". $data["numgameservers"]["active"] ."/". $data["numgameservers"]["all"] ." responded):\n";
	print_r ($data["gameservers"]);
	echo "</pre></body></html>";
}
?>